src/diffusers/pipelines/pipeline_utils.py

```diff
@@ -722,6 +722,23 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
             )
         else:
             cached_folder = pretrained_model_name_or_path
+            filenames = []
+            for _, _, files in os.walk(cached_folder):
+                for file in files:
+                    filenames.append(os.path.basename(file))
+
+            model_filenames, variant_filenames = variant_compatible_siblings(filenames, variant=variant)
+            # The variant filenames can have the legacy sharding checkpoint format that we check and throw
+            # a deprecation warning if detected.
+            if len(variant_filenames) > 0 and _check_legacy_sharding_format(filenames, variant):
+                deprecation_message = f"This serialization format is now deprecated to standardize the serialization format between `transformers` and `diffusers`. We recommend you to remove the existing files associated with the current variant ({variant}) and re-obtain them by running a `save_pretrained()`."
+                deprecate("legacy_sharded_ckpts_with_variant", "1.0.0", deprecation_message, standard_warn=False)
+            if len(variant_filenames) == 0 and variant is not None:
+                error_message = (
+                    f"You are trying to load the model files of the `variant={variant}`, but no such modeling files are available."
+                    f" Available ones are: {model_filenames}."
+                )
+                raise ValueError(error_message)
```

Comment:
this code block here it seems all it does is:
1. throw an error when `len(variant_filenames) == 0` and `variant is not None`
2. warn about the legacy format 

In that case, to keep things simple, I think it is sufficient to add the check after `_identify_model_variant(...)` line a little bit below here on line 725/752; we can just run the check on its output `model_variants`

in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

another note is for legacy format, here we can just add a warning, no need to deprecate as it was never supported

Reply:
> in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

I just replaced the old `_identify_model_variant()` from https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140 and ran `pytest tests/pipelines/stable_diffusion/test_stable_diffusion.py -k "test_loading_with_variants"` and it error out. With the fixed method, the error was gone. 

Agree with rest of the points. 

The mistake wasn't caught because we didn't have enough tests to catch it, IMO.

Reply:
> in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

I just replaced the old `_identify_model_variant()` from https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140 and ran `pytest tests/pipelines/stable_diffusion/test_stable_diffusion.py -k "test_loading_with_variants"` and it error out. With the fixed method, the error was gone. 

Agree with rest of the points. 

The mistake wasn't caught because we didn't have enough tests to catch it, IMO.

Reply:
> in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

I just replaced the old `_identify_model_variant()` from https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140 and ran `pytest tests/pipelines/stable_diffusion/test_stable_diffusion.py -k "test_loading_with_variants"` and it error out. With the fixed method, the error was gone. 

Agree with rest of the points. 

The mistake wasn't caught because we didn't have enough tests to catch it, IMO.

Reply:
> in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

I just replaced the old `_identify_model_variant()` from https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140 and ran `pytest tests/pipelines/stable_diffusion/test_stable_diffusion.py -k "test_loading_with_variants"` and it error out. With the fixed method, the error was gone. 

Agree with rest of the points. 

The mistake wasn't caught because we didn't have enough tests to catch it, IMO.

Reply:
> in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

I just replaced the old `_identify_model_variant()` from https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140 and ran `pytest tests/pipelines/stable_diffusion/test_stable_diffusion.py -k "test_loading_with_variants"` and it error out. With the fixed method, the error was gone. 

Agree with rest of the points. 

The mistake wasn't caught because we didn't have enough tests to catch it, IMO.

Reply:
> in fact, adding the check here wouldn't have caught the mistake you made here https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140

I just replaced the old `_identify_model_variant()` from https://github.com/huggingface/diffusers/pull/8967#discussion_r1727809140 and ran `pytest tests/pipelines/stable_diffusion/test_stable_diffusion.py -k "test_loading_with_variants"` and it error out. With the fixed method, the error was gone. 

Agree with rest of the points. 

The mistake wasn't caught because we didn't have enough tests to catch it, IMO.